### PR TITLE
Deduplication offset period errors handled in Sandbox-on-X [DPP-872]

### PR DESCRIPTION
**Pull request by tudor-da**
Deduplication offset period errors handled in Sandbox-on-X
* Adds `InternalUnsupportedOperation` error category

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/error/src/main/scala/com/daml/error/ErrorCategory.scala [hunk 1]
```diff
@@ -332,8 +333,28 @@ object ErrorCategory {
       )
       with ErrorCategory
 
-  implicit val orderingErrorType: Ordering[ErrorCategory] = Ordering.by[ErrorCategory, Int](_.rank)
+  @Description(
+    """This error category is used to signal that an unimplemented code-path has been triggered by a client or participant operator request.
+      |Such an error can arise expectedly if the operator submits an admin request against a participant that does not, partly or fully,
+      |support the targeted feature. This error can also arise when a user request ends up leveraging, unexpectedly, an unsupported feature."""
```
Comment (meiersi-da) [inline ledger/error/src/main/scala/com/daml/error/ErrorCategory.scala:339]:
```suggestion
```
I'd be inclined to remove this explanation. It's a bit fuzzy; and I don't know how to make it better. I believe the most important information is the resolution below.

File: ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala [hunk 2]
```diff
@@ -23,6 +23,18 @@ import java.time.{Duration, Instant}
   "Errors raised by or forwarded by the Ledger API."
 )
 object LedgerApiErrors extends LedgerApiErrorGroup {
+  object UnsupportedFeature
+      extends ErrorCode(
+        id = "UNSUPPORTED_FEATURE",
```
Comment (meiersi-da) [inline ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala:28]:
```suggestion
        id = "UNSUPPORTED_OPERATION",
```
align to internal name

File: ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala [hunk 3]
```diff
@@ -23,6 +23,18 @@ import java.time.{Duration, Instant}
...
+        ErrorCategory.InternalUnsupportedOperation,
+      ) {
+
+    case class Reject(unsupportedFeature: String)(implicit errorLogger: ContextualizedErrorLogger)
+        extends LoggingTransactionErrorImpl(
+          cause = s"The request exercised an unsupported ledger feature: $unsupportedFeature"
```
Comment (meiersi-da) [inline ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala:34]:
```suggestion
          cause = s"The request exercised an unsupported operation: $unsupportedFeature"
```

Make it wider so we don't need to introduce more kinds of unsupported operation errors.

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/domain/Rejection.scala [hunk 4]
```diff
@@ -64,6 +64,14 @@ private[sandbox] object Rejection {
       .rpcStatus(completionInfo.submissionId)
   }
 
+  final case class OffsetDeduplicationPeriodUnsupported(completionInfo: CompletionInfo)(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ) extends Rejection {
+    override def toStatus: Status = LedgerApiErrors.UnsupportedFeature
+      .Reject("offset-type deduplication period")
```
Comment (meiersi-da) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/domain/Rejection.scala:71]: I'd wager that the error message is quite hard to understand, as "offset-type deduplication period" is not really an operation. I'd expect there to be a verb or a sentence for describing an operation.

File: ledger/error/src/main/scala/com/daml/error/ErrorCategory.scala [hunk 5]
```diff
@@ -332,8 +333,26 @@ object ErrorCategory {
...
       with ErrorCategory
 
-  implicit val orderingErrorType: Ordering[ErrorCategory] = Ordering.by[ErrorCategory, Int](_.rank)
+  @Description(
+    """This error category is used to signal that an unimplemented code-path has been triggered by a client or participant operator request."""
+  )
+  @RetryStrategy("""Errors in this category are non-retryable.""")
+  @Resolution(
+    """This error is caused by a ledger-level misconfiguration or by an implementation bug.
+      |Resolution requires ledger operator intervention or vendor support."""
```
Comment (meiersi-da) [inline ledger/error/src/main/scala/com/daml/error/ErrorCategory.scala:342]:
```suggestion
      |Resolution requires participant operator intervention."""
```

The operator can then escalate to the vendor if required.

File: ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala [hunk 6]
```diff
@@ -27,6 +27,25 @@ object LedgerApiErrors extends LedgerApiErrorGroup {
 
   val EarliestOffsetMetadataKey = "earliest_offset"
 
+  @Explanation(
+    """This error category is used to signal that an unimplemented code-path has been triggered by a client or participant operator request."""
+  )
+  @Resolution(
+    """This error is caused by a ledger-level misconfiguration or by an implementation bug.
```
Comment (meiersi-da) [inline ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala:34]:
```suggestion
    """This error is caused by a participant node misconfiguration or by an implementation bug.
```

File: ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala [hunk 7]
```diff
@@ -27,6 +27,25 @@ object LedgerApiErrors extends LedgerApiErrorGroup {
...
+      |Resolution requires ledger operator intervention or vendor support."""
```
Comment (meiersi-da) [inline ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala:35]:
```suggestion
      |Resolution requires participant operator intervention."""
```

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/domain/Rejection.scala [hunk 8]
```diff
@@ -64,6 +64,14 @@ private[sandbox] object Rejection {
       .rpcStatus(completionInfo.submissionId)
   }
 
+  final case class OffsetDeduplicationPeriodUnsupported(completionInfo: CompletionInfo)(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ) extends Rejection {
+    override def toStatus: Status = LedgerApiErrors.UnsupportedOperation
+      .Reject("Command deduplication period of type offset is not supported.")
```
Comment (meiersi-da) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/domain/Rejection.scala:71]:
```suggestion
      .Reject("command deduplication with periods specified using offsets")
```

which results in the message "The request exercised an unsupported operation: command deduplication with periods specified using offsets"

**Review by meiersi-da — APPROVED** [review]
Thanks for the review pass @tudor-da . I've added some more reformulation suggestions. The important point when writing these is to consider what understanding and behaviour they foster in the reader. They should be clear, proper English, and encourage the reader to fix the problem him/herself.

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/domain/Rejection.scala (hunk 4 above)
Comment (tudor-da) [reply ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/domain/Rejection.scala]: rephrased

**Pull request merged**